proto/provenance/exchange/v1/tx.proto

```diff
@@ -0,0 +1,367 @@
+message MsgMarketWithdrawResponse {}
+
+// MsgMarketUpdateDetailsRequest is a request message for the MarketUpdateDetails endpoint.
+message MsgMarketUpdateDetailsRequest {
+  option (cosmos.msg.v1.signer) = "admin";
+
+  // admin is the account with "update" permission requesting this change.
+  string admin = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // market_id is the numerical identifier of the market to update required attributes for.
+  uint32 market_id = 2;
+
+  // market_details is some information about this market.
+  MarketDetails market_details = 3 [(gogoproto.nullable) = false];
+}
+
+// MsgMarketUpdateDetailsResponse is a response message for the MarketUpdateDetails endpoint.
+message MsgMarketUpdateDetailsResponse {}
+
+// MsgMarketUpdateEnabledRequest is a request message for the MarketUpdateEnabled endpoint.
+message MsgMarketUpdateEnabledRequest {
+  option (cosmos.msg.v1.signer) = "admin";
+
+  // admin is the account with "update" permission requesting this change.
+  string admin = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // market_id is the numerical identifier of the market to update required attributes for.
+  uint32 market_id = 2;
+
+  // accepting_orders is whether this market is allowing orders to be created for it.
+  bool accepting_orders = 3;
+}
+
+// MsgMarketUpdateEnabledResponse is a response message for the MarketUpdateEnabled endpoint.
+message MsgMarketUpdateEnabledResponse {}
+
+// MsgMarketUpdateUserSettleRequest is a request message for the MarketUpdateUserSettle endpoint.
+message MsgMarketUpdateUserSettleRequest {
+  option (cosmos.msg.v1.signer) = "admin";
+
+  // admin is the account with "update" permission requesting this change.
+  string admin = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // market_id is the numerical identifier of the market to update required attributes for.
+  uint32 market_id = 2;
+
+  // allow_user_settlement is whether this market allows users to initiate their own settlements.
+  // For example, the FillBids and FillAsks endpoints are available if and only if this is true.
+  // The MarketSettle endpoint is only available to market actors regardless of the value of this field.
+  bool allow_user_settlement = 3;
+}
+
+// MsgMarketUpdateUserSettleResponse is a response message for the MarketUpdateUserSettle endpoint.
+message MsgMarketUpdateUserSettleResponse {}
+
+// MsgMarketManagePermissionsRequest is a request message for the MarketManagePermissions endpoint.
+message MsgMarketManagePermissionsRequest {
+  option (cosmos.msg.v1.signer) = "admin";
+
+  // admin is the account with "permissions" permission requesting this change.
+  string admin = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // market_id is the numerical identifier of the market to update required attributes for.
+  uint32 market_id = 2;
+
+  // revoke_all are addresses that should have all their permissions revoked.
+  repeated string revoke_all = 3;
```

Comment:
Is there a difference between `revoke_all` and adding all permissions to `to_revoke`? It would get rid of a lot of extra validation and logic if we only used one.

Reply:
In `to_revoke` you can only include permissions that they have. So you'd have to look up their permissions before you could revoke them. Most of the time, when permissions need to be revoked, it needs to be all their permissions, and its often time critical. So it's good to make it easy to do that.

If I got rid of the `to_revoke` list (and just had `revoke_all` and `to_grant`), it becomes painful to revoke just a single permission since, again, you'd have to look up their existing permissions first.